### PR TITLE
Add/Improve unit tests for compaction with histogram

### PR DESCRIPTION
**Pull request by codesome**
NOTE: this is for sparsehistogram branch.

This PR takes care of the following from #11171

	- [ ] Compaction
		- [ ] normal compaction works (head block compaction and merging of 2 non-overlapping blocks)
			- [ ] Only histograms in a series (with other float series)
			- [ ] Mix of histograms and float in a series in the same block (with other float series)
			- [ ] One block having histogram series and the other block having float series
			- [ ] Multiple blocks having a mix of float and histograms for the same series
		- [ ] Vertical compaction works
			- [ ] A series with only histograms overlaps with a series with only histograms
			- [ ] A series with only histograms overlaps with a series with histograms and float
			- [ ] A series with only histograms overlaps with a series with only float
			- [ ] A series with histograms and float overlaps with a series with only float

The PR
* Improves the existing head compaction test
* Removed an unnecessary test that was testing block compaction.
* Extend existing query on blocks test to also do compaction test

### CODE REVIEW

File: tsdb/compact_test.go [hunk 1]
```diff
@@ -1298,29 +1299,79 @@ func TestDeleteCompactionBlockAfterFailedReload(t *testing.T) {
...
 	})
 
-	require.NoError(t, head.Init(0))
-	app := head.Appender(context.Background())
+	minute := func(m int) int64 { return int64(m) * time.Minute.Milliseconds() }
+	ctx := context.Background()
+	appendHistogram := func(lbls labels.Labels, from, to int, h *histogram.Histogram, exp *[]tsdbutil.Sample) {
+		t.Helper()
+		app := head.Appender(ctx)
+
```
Comment (beorn7) [inline tsdb/compact_test.go:1312]: Nit: Blank line here but not below in appendFloat.

File: tsdb/compact_test.go [hunk 2]
```diff
@@ -1298,29 +1299,79 @@ func TestDeleteCompactionBlockAfterFailedReload(t *testing.T) {
...
-	// Ingest samples.
-	numHistograms := 120 * 4
-	timeStep := DefaultBlockDuration / int64(numHistograms)
-	expHists := make([]timedHistogram, 0, numHistograms)
-	l := labels.Labels{{Name: "a", Value: "b"}}
-	for i, h := range GenerateTestHistograms(numHistograms) {
-		_, err := app.AppendHistogram(0, l, int64(i)*timeStep, h)
-		require.NoError(t, err)
-		expHists = append(expHists, timedHistogram{int64(i) * timeStep, h})
+	// Append all scenario of histograms.
```
Comment (beorn7) [inline tsdb/compact_test.go:1332]: I'm confused by this doc comment. Maybe "Append floats and histograms for all scenarios."? Not sure what you want to express. Maybe no comment is needed?

File: tsdb/db_test.go [hunk 3]
```diff
@@ -3999,11 +4016,13 @@ func TestQueryHistogramFromBlocks(t *testing.T) {
 		)
 	})
 
-	t.Run("overlapping blocks with only histograms", func(t *testing.T) {
+	t.Run("serial blocks with only histograms and only float in a series", func(t *testing.T) {
```
Comment (beorn7) [inline tsdb/db_test.go:4019]: The description confuses me. Maybe "serial blocks: histograms, floats, histograms" or similar?

**Review by beorn7 — APPROVED** [review]
Just formatting and naming nits.

File: tsdb/db_test.go (hunk 3 above)
Comment (codesome) [reply tsdb/db_test.go:4019]: I meant that a block has either histograms or float and not both

**Comment by codesome** [comment]
Even a rerun isn't helping. The failure happened before as well. The test passes locally and the error logs that say failed to open meta.json also appear in the passed test, so I don't think that is the problem. I will see what is up.

**Comment by codesome** [comment]
Closing in favour of https://github.com/prometheus/prometheus/pull/11342 and https://github.com/prometheus/prometheus/pull/11343. I will investigate the error in those PR.

**Comment by codesome** [comment]
Error is from changes in db_test.go

**Pull request closed without merging**